### PR TITLE
CNS-800: Verify the token denom in Tx

### PR DESCRIPTION
**Pull request by shleikes**

### CODE REVIEW

File: x/dualstaking/client/cli/tx_delegate.go [hunk 1]
```diff
@@ -50,6 +52,11 @@ func CmdDelegate() *cobra.Command {
 				argChainID,
 				argAmount,
 			)
+
+			if msg.Amount.Denom != commontypes.TokenDenom {
```
Comment (Yaroms) [inline x/dualstaking/client/cli/tx_delegate.go:56]: token denom validation needs to be in the message server, not in the CLI.
we dont want to use the commontypes.tokenDenom

Comment (Yaroms) [reply x/dualstaking/client/cli/tx_delegate.go:56]: this is also the common practice in cosmos

File: x/dualstaking/types/errors.go [hunk 2]
```diff
@@ -13,4 +13,5 @@ var (
 	ErrBadDelegationAmount       = sdkerrors.Register(ModuleName, 1003, "invalid delegation amount")
 	ErrUnbondingInProgress       = sdkerrors.Register(ModuleName, 1004, "unbonding already exists (same block)")
 	ErrCalculatingProviderReward = sdkerrors.Register(ModuleName, 1005, "provider reward calculation failed")
+	ErrWrongDenom                = sdkerrors.Register(ModuleName, 1006, "amount has wrong denomanator")
```
Comment (oren-lava) [inline x/dualstaking/types/errors.go:16]: denominator

Comment (oren-lava) [reply x/dualstaking/types/errors.go:16]: create an error message in the common module (or utils) and use a single message for all modules (I can see there is a different message in the plans module)

**Review by oren-lava — CHANGES_REQUESTED** [review]

File: x/dualstaking/client/cli/tx_delegate.go (hunk 1 above)
Comment (shleikes) [reply x/dualstaking/client/cli/tx_delegate.go]: Fixed in fcee22d6fcfd4ffacef8ec0181f5ee04ed1495c7

File: x/dualstaking/types/errors.go (hunk 2 above)
Comment (shleikes) [reply x/dualstaking/types/errors.go]: Fixed in 1697968c86cc5c83a771ee9803e811b3a1b1a9dd

File: x/pairing/client/cli/tx_stake_provider.go [hunk 3]
```diff
@@ -100,7 +100,8 @@ func CmdStakeProvider() *cobra.Command {
 				commission,
 			)
 
-			if msg.DelegateLimit.Denom != commontypes.TokenDenom {
+			if msg.DelegateLimit.Denom != commontypes.TokenDenom ||
```
Comment (Yaroms) [inline x/pairing/client/cli/tx_stake_provider.go:103]: move this to msg_server

Comment (shleikes) [reply x/pairing/client/cli/tx_stake_provider.go]: Fixed in 9140f41f2dc52a2dfd2902c4170f53baed67704b

**Review by oren-lava — APPROVED** [review]

File: x/dualstaking/keeper/delegate.go [hunk 4]
```diff
@@ -283,10 +272,13 @@ func (k Keeper) delegate(ctx sdk.Context, delegator, provider, chainID string, a
...
 
-	if err := validateCoins(amount); err != nil {
-		return err
-	} else if amount.IsZero() {
-		return nil
+	if err := utils.ValidateCoins(ctx, k.stakingKeeper.BondDenom(ctx), amount, false); err != nil {
+		return utils.LavaFormatWarning("failed to delegate: coin validation failed", err,
+			utils.Attribute{Key: "delegator", Value: delegator},
+			utils.Attribute{Key: "provider", Value: provider},
+			utils.Attribute{Key: "amount", Value: amount.String()},
```
Comment (Yaroms) [inline x/dualstaking/keeper/delegate.go:279]: the amount is already in the err from validate coin

File: x/dualstaking/keeper/delegate.go [hunk 5]
```diff
@@ -330,10 +322,12 @@ func (k Keeper) Redelegate(ctx sdk.Context, delegator, from, to, fromChainID, to
...
 
-	if err := validateCoins(amount); err != nil {
-		return err
-	} else if amount.IsZero() {
-		return nil
+	if err := utils.ValidateCoins(ctx, k.stakingKeeper.BondDenom(ctx), amount, false); err != nil {
+		return utils.LavaFormatWarning("failed to redelegate: coin validation failed", err,
+			utils.Attribute{Key: "delegator", Value: delegator},
+			utils.Attribute{Key: "provider", Value: to},
+			utils.Attribute{Key: "amount", Value: amount.String()},
```
Comment (Yaroms) [inline x/dualstaking/keeper/delegate.go:329]: same

File: x/dualstaking/keeper/delegate.go [hunk 6]
```diff
@@ -382,10 +376,12 @@ func (k Keeper) unbond(ctx sdk.Context, delegator, provider, chainID string, amo
...
 
-	if err := validateCoins(amount); err != nil {
-		return err
-	} else if amount.IsZero() {
-		return nil
+	if err := utils.ValidateCoins(ctx, k.stakingKeeper.BondDenom(ctx), amount, false); err != nil {
+		return utils.LavaFormatWarning("failed to unbond: coin validation failed", err,
+			utils.Attribute{Key: "delegator", Value: delegator},
+			utils.Attribute{Key: "provider", Value: provider},
+			utils.Attribute{Key: "amount", Value: amount.String()},
```
Comment (Yaroms) [inline x/dualstaking/keeper/delegate.go:383]: same

File: x/dualstaking/keeper/delegate.go (hunk 4 above)
Comment (shleikes) [reply x/dualstaking/keeper/delegate.go]: Fixed in 4a9bd33080264c1f8f20a13cfead74ce23227bca

File: x/dualstaking/keeper/delegate.go (hunk 5 above)
Comment (shleikes) [reply x/dualstaking/keeper/delegate.go]: Fixed in 4a9bd33080264c1f8f20a13cfead74ce23227bca

File: x/dualstaking/keeper/delegate.go (hunk 6 above)
Comment (shleikes) [reply x/dualstaking/keeper/delegate.go]: Fixed in 4a9bd33080264c1f8f20a13cfead74ce23227bca

**Review by Yaroms — APPROVED** [review]

**Pull request merged**